### PR TITLE
Final Preparations for 330 - Sherif

### PR DESCRIPTION
**Pull request by zasherif**

### CODE REVIEW

File: docs/design/dp_programability.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+Table: Endpoint table of S00 and S01
+
+### Summary of Workflows
+
+The following table summarizes the primary control-plane workflows as well as the table to be populated during each operation.
+
+| User API/Events | Control-Plane Algorithm                                                    | Data-plane RPCs                                                                                                                                      | Data-plane ebpf map                                                         |
+|-----------------|----------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------|
+| Create VPC      | Allocate dividers for the VPC                                       | call update_vpc on all dividers of the VPC (optional)                                                                                         | Popoulate VPC map (transit XDP program)                                     |
```
Comment (phudtran) [inline docs/design/dp_programability.md]: call **update_network** on all dividers

**Populate** VPC map

File: docs/design/mp_overview.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+implementation the callback is translated into one of multiple modes.
+In all of these modes, the operator developer only focuses on the
+workflow through calling the callback function without worrying about
+the details of the communication implementation.
+
+1. **Through the API Server(s)**: In this mode, operators communication is
+   implicit with the help of the API server. As one operator makes the
+   callback to invoke the next step of the object pipeline, the
+   only changes the object state and update the object rhough the API
+   server. As the next Operator in the pipeline handles the object in
```
Comment (phudtran) [inline docs/design/mp_overview.md]: For this part,

I think you mean:

", the **operator** only changes the object state and **updates** the object **through** the API Server

File: docs/design/mp_overview.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+   the new state, communication happens implicitly. The drawback of
+   this approach is the loading on the API server, the latency
+   involved to evolve the object's pipeline. The advantage is that the
```
Comment (phudtran) [inline docs/design/mp_overview.md]: Here I think we should say:

The drawback of this approach is the **load** on the API server, **and** the latency involved to evolve the object's pipeline.

File: docs/design/mp_overview.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+   calling operators must handle failure scenarios in such case. The
+   drawback of this approach is the complexity of handling failure
+   scenarios. Also, as the intermediate states of the objects are not
+   updates to the API server, it becomes harder to troubleshoot
+   problems in the object-pipeline. This approach __may__ have latency
+   advantage, especially if the API server communication is a
+   bottlneck. However, if the API server and kubernetes storage scales
+   horizontally latency of the this approach and the second approach
+   may be comparable. In that case, this approac only adds
+   implementation complexitie that are hard to diagnose.
```
Comment (phudtran) [inline docs/design/mp_overview.md]: In that case, this **approach** only adds implementation **complexities** that are hard to diagnose.

File: docs/design/mp_overview.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+1. **Simple Function Calls**: In this mode, all operators are deployed
+   into a single (horizontally scalable) operator. The functional
+   implementation of the operators is still seperated through
+   call-back interfaces. But the callbacks are implemented as simply
+   function calls. This approach is the simplist to develop, test, and
```
Comment (phudtran) [inline docs/design/mp_overview.md]: This approach is the **simplest** to develop

File: docs/design/mp_overview.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+troubleshooting becomes less of a concern in the simple function calls
+mode because state transitions is within the same process with unified
+logging, and metering.
+
+### Local object store (LMDB) and Resumable Workflow
+
+Each Operator will have a local K/V store implemented with
+[LMDB](https://symas.com/lmdb/). Each Operator stores its Stateful
+Object in this store when it reaches the state "Provisioned" and
+delete it when the object is recycled (deleted). The objects are
```
Comment (phudtran) [inline docs/design/mp_overview.md]: **deletes** it

File: docs/design/mp_overview.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+serving stalled objects in the API server (in Init or recorded
+intermediate state) utilizing the data of the provisioned objects in
+the local store. In this regard, the local store serves two purposes:
+
+1. It caches the final object's data into its operator, hence reduces
+   the need to frequently get the object through the API server.
+1. The store accelerates the starting time of the operators after a
+   failure.
+
+#### Extensiton to an extenrnal store
```
Comment (phudtran) [inline docs/design/mp_overview.md]: **Extension** to an **external** store

File: docs/design/dp_datamodel.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
+
+Mizar's data-plane adopts a simple data model that is essential to extend its functionality. The data-model is independant of the API definitions of the management-plane. Typically, different management plane implementations that encorporated Mizar as its data-plane has a shim layer that translates management-plane data-model into Mizar's constructs. The model consists of endpoint, network, and VPC constructs.
```
Comment (phudtran) [inline docs/design/dp_datamodel.md]: Typically, different management plane implementations that **incorporated** Mizar

File: README.md [hunk 9]
```diff
@@ -11,8 +11,49 @@ Mizar network has the following advantages:
...
+Mizar's [data-plane](docs/design/dp_overview.md) provides high performance and extensible packet processing pipeline and functions that achieve Mizar's functional, scale, and performance goals. Mizar's [management-plane](docs/design/mp_overview.md) programs the data-plane by translating typical networking related APIs and resources to Mizar specific configuration. The programmability of the data-plane involves loading and unloading network functions at various stages of the packet processing pipeline. Mizar management-plane can support different networking technologies. The data-plane has been successfuly integrated as well in other implementations of management-planes.
+
+## Why Mizar is different?
+
+Unlike traditional networking solutions, Mizar relies on the natural partitioning of a cloud network to scale. Mizar simplifies the programming of data-plane to scale by flexible in-network processing, compared to flow-based programming models. As it primarily targets use cases of cloud-networking among virtual machines and containers, Mizar reduces the control-plane overhead of several routing and switching protocols within a cloud environment (e.g., L2 learning, ARP, BGP, etc.).
```
[the remark this answers is not in the document]
Comment (zasherif) [reply README.md]: OK

File: docs/design/dp_extensibility.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
+Mizar facilitates the development of **Multi-tenant** networking functions at various levels of the stack!
```
[the remark this answers is not in the document]
Comment (zasherif) [reply docs/design/dp_extensibility.md]: OK

File: docs/design/dp_programability.md (hunk 1 above)
Comment (zasherif) [reply docs/design/dp_programability.md]: OK

File: docs/design/mp_overview.md (hunk 2 above)
Comment (zasherif) [reply docs/design/mp_overview.md]: ok

File: docs/design/mp_overview.md (hunk 3 above)
Comment (zasherif) [reply docs/design/mp_overview.md]: ok

File: docs/design/mp_overview.md (hunk 4 above)
Comment (zasherif) [reply docs/design/mp_overview.md]: OK

File: docs/design/mp_overview.md (hunk 5 above)
Comment (zasherif) [reply docs/design/mp_overview.md]: ok

File: docs/design/mp_overview.md (hunk 6 above)
Comment (zasherif) [reply docs/design/mp_overview.md]: ok

File: docs/design/mp_overview.md (hunk 7 above)
Comment (zasherif) [reply docs/design/mp_overview.md]: ok

**Pull request merged**